### PR TITLE
Added page "Community Reviews"

### PR DESCRIPTION
**Pull request by webmozart**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | - |

This is the WIP draft for the community review process. Feedback welcome :)

### CODE REVIEW

File: contributing/community/reviews.rst [hunk 1]
```diff
@@ -0,0 +1,190 @@
+Community Reviews
+=================
+
+Symfony is an open-source project driven by a large community. If you don't feel
+ready to contribute code or patches, reviewing issues and pull requests (PRs)
+can be a great start to get involved and give back.
+
+Why Reviewing is Important
```
Comment (xabbuh) [inline contributing/community/reviews.rst]: Is

File: contributing/community/reviews.rst [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+Note that **anyone who has some basic familiarity with Symfony and PHP can
+review bug reports and pull requests**. You don't need to be an expert to help.
+
+Be Constructive
+---------------
+
+Before you begin, remember that you are looking at the result of someone else's
+hard work. A good review post thanks the contributor for their work, identifies
+what was done well, identifies what should be improved, and suggests a next
```
Comment (xabbuh) [inline contributing/community/reviews.rst]: The comma before and should be removed.

File: contributing/community/reviews.rst [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+The Bug Report Review Process
+-----------------------------
+
+A good way to get started with reviewing is to pick a bug report from the
+`bug reports in need of review`_.
+
+The steps for the review are:
+
+1. **Is the Report Complete?**
```
Comment (xabbuh) [inline contributing/community/reviews.rst]: Please use `#.` rather than concrete numbers.

File: contributing/community/reviews.rst [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+   Good bug reports contain a link to a fork of the `Symfony Standard Edition`_
+   (the "reproduction project") that reproduces the bug. If it doesn't, the
+   report should at least contain enough information and code samples to
+   reproduce the bug.
```
Comment (xabbuh) [inline contributing/community/reviews.rst]: Should we put focus on the Symfony SE fork first? Imo many bugs are easily reproducable by describing some simple steps or by giving some small example code snippet. The thing is, we should make it as easy as possible to contribute and forking the Standard Edition is obviously more work than writing text that provides the same details.

So I think we should keep this paragraph, but just resort a bit which approach we put focus on.

File: contributing/community/reviews.rst [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+   Download the reproduction project and test whether the bug can be reproduced
+   on your system. If the reporter did not provide a reproduction project,
+   create one by forking_ the `Symfony Standard Edition`_. Reproduce the bug
+   with the instructions given by the reporter.
+
+3. **Update the Issue Status**
+
+   At last, add a comment and update the status of the bug report. **Thank the
+   reporter for reporting the bug**. Include the line ``Status: <status>`` in
```
Comment (xabbuh) [inline contributing/community/reviews.rst]: Do we need to explain that this format is necessary as it will be used to automatically change the state of the issue? I'm not sure if everyone understands this from just reading these lines and they may be confused by not finding any button to change the state.

File: contributing/community/reviews.rst [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+It is okay to do partial reviews. If you do a partial review, post how far you
+got and leave the PR in "Needs Review" state.
+
+Pick a pull request from the `PRs in need of review`_ and follow these steps:
+
+1. **Is the PR Complete**?
+
+   Every pull request must contain a header that gives some basic information
+   about the PR. You can find the template for that header in the
+   `Contribution Guidelines`_.
```
Comment (xabbuh) [inline contributing/community/reviews.rst]: IIRC, we have an anchor for the section that contains the template and and can just link to this directly. Otherwise, we should at least use the `doc` role here for internal links.

Comment (xabbuh) [inline contributing/community/reviews.rst]: Ah, no we don't. But we should add one just before the "Make a Pull Request" headline in `/contributing/code/patches.rst`:

```rst
.. _contributing-code-pull-request:

Make a Pull Request```

and then refer to this anchor:

```rst
[...] in the :ref:`Contribution Guidelines <contributing-code-pull-request>`.
```

File: contributing/community/reviews.rst [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+2. **Is the Base Branch Correct?**
+
+   GitHub displays the branch that a PR is based on below the title of the
+   pull request. Is that branch correct?
+
+   * Bugs should be fixed in the oldest, maintained version that contains the
+     bug. Check `Symfony's Release Schedule`_ to find the oldest currently
```
Comment (xabbuh) [inline contributing/community/reviews.rst]: Please use the `doc` role for internal links.

File: contributing/community/reviews.rst [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+     supported version.
+
+   * New features should always be added to the current development version.
+     Check the `Symfony Roadmap`_ to find the current development version.
```
Comment (xabbuh) [inline contributing/community/reviews.rst]: We should explain here that you need to differ between BC breaking changes and BC compatible changes given that we have development branches for 2.x and 3.x, shouldn't we?

File: contributing/community/reviews.rst [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+   * Does the PR contain automated tests? Do those tests cover all relevant
+     edge cases?
+   * Does the PR contain sufficient comments to easily understand its code?
+   * Does the code break backwards compatibility? If yes, does the PR header say
+     so?
+   * Does the PR contain deprecations? If yes, does the PR header say so? Does
+     the code contain ``trigger_error()`` statements for all deprecated
+     features?
+   * Are all deprecations and backwards compatibility breaks documented in the
+     latest UPGRADE-X.X.md file? Do those explanations contain "Before"/"After"
```
Comment (xabbuh) [inline contributing/community/reviews.rst]: [...] "Before"/"After" examples [...]

**Comment by xabbuh** [comment]
Thanks Bernhard. I left some small comments, but overall I think this is a great document to help people contributing. :+1:

File: contributing/community/reviews.rst [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+  backwards compatibility.
+
+Note that **anyone who has some basic familiarity with Symfony and PHP can
+review bug reports and pull requests**. You don't need to be an expert to help.
+
+Be Constructive
+---------------
+
+Before you begin, remember that you are looking at the result of someone else's
+hard work. A good review post thanks the contributor for their work, identifies
```
Comment (lsmith77) [inline contributing/community/reviews.rst]: s/post/comment

File: contributing/community/reviews.rst [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+Community reviews are essential for the development of the Symfony framework,
+since there are many more pull requests and bug reports than there are members
+in the Symfony core team to review, fix and merge them.
+
+On the `Symfony issue tracker`_, you can find many items in a "Needs Review"
+status:
+
+* **Bug Reports**: Bug reports primarily need to be checked for completeness.
+  Is any important information missing? Can the bug easily be reproduced?
```
Comment (cordoval) [inline contributing/community/reviews.rst]: be easily reproduced?

File: contributing/community/reviews.rst (hunk 4 above)
Comment (lsmith77) [inline contributing/community/reviews.rst]: yeah ... i think we should classify the bugs a bit more to suggest when an SE fork is the sensible way to provide a reproducible test case .. in the end for a simple bug a description will be more efficient than having to download and run an SE fork (though what would be awesome is if platform.sh would provide some infrastructure to quickly spin up such forks for core devs).

File: contributing/community/reviews.rst [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+   Good bug reports contain a link to a fork of the `Symfony Standard Edition`_
+   (the "reproduction project") that reproduces the bug. If it doesn't, the
+   report should at least contain enough information and code samples to
+   reproduce the bug.
+
+2. **Reproduce the Bug**
+
+   Download the reproduction project and test whether the bug can be reproduced
+   on your system. If the reporter did not provide a reproduction project,
+   create one by forking_ the `Symfony Standard Edition`_. Reproduce the bug
```
Comment (lsmith77) [inline contributing/community/reviews.rst]: see above about the SE fork

File: contributing/community/reviews.rst (hunk 5 above)
Comment (lsmith77) [inline contributing/community/reviews.rst]: +1 for explaining that this will trigger a bot to update the labels

File: contributing/community/reviews.rst [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+    Here is a sample comment for a bug report that could be reproduced:
+
+    .. code-block:: text
+
+        Thank you @weaverryan for creating this bug report! This indeed looks
+        like a bug. I reproduced the bug in the "kernel-bug" branch of
+        https://github.com/webmozart/symfony-standard.
+
+        Status: Reviewed
```
Comment (cordoval) [inline contributing/community/reviews.rst]: This sounds a bit redundant as anyone can set a status?

File: contributing/community/reviews.rst [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+
+   At last, add a comment and update the status of the bug report. **Thank the
+   reporter for reporting the bug**. Include the line ``Status: <status>`` in
+   your comment to update the status to one of the following:
+
+   **Needs Work** If the bug does not contain enough information to be
+   reproduced, explain what information is missing and move the report to this
+   status.
+
+   **Works for me** If the bug *does* contain enough information to be
```
Comment (lsmith77) [inline contributing/community/reviews.rst]: should be have 2 states for this? one for not reproducible and one for new feature?

File: contributing/community/reviews.rst [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+
+.. topic:: Example
+
+    Here is a sample comment for a bug report that could be reproduced:
+
+    .. code-block:: text
+
+        Thank you @weaverryan for creating this bug report! This indeed looks
+        like a bug. I reproduced the bug in the "kernel-bug" branch of
+        https://github.com/webmozart/symfony-standard.
```
Comment (lsmith77) [inline contributing/community/reviews.rst]: should'nt the link just be ``https://github.com/webmozart/symfony-standard/tree/kernel-bug`` ?

File: contributing/community/reviews.rst [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+The Pull Request Review Process
+-------------------------------
+
+The process for reviewing pull requests (PRs) is similar to doing a review of a
+bug report. Reviews of pull requests usually take a little longer since you need
+to understand the functionality that has been fixed or implemented and then find
+out whether the implementation is complete.
+
+It is okay to do partial reviews. If you do a partial review, post how far you
```
Comment (lsmith77) [inline contributing/community/reviews.rst]: s/post/comment

File: contributing/community/reviews.rst (hunk 7 above)
Comment (lsmith77) [inline contributing/community/reviews.rst]: We might need to expand this .. possibly including an example .. maybe we should even explicitly state the version and ensure we manually update it.

File: contributing/community/reviews.rst (hunk 8 above)
Comment (lsmith77) [inline contributing/community/reviews.rst]: +1

File: contributing/community/reviews.rst [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+
+3. **Reproduce the Problem**
+
+   Read the issue that the pull request is supposed to fix. Reproduce the
+   problem on a clean `Symfony Standard Edition`_ project and try to understand
+   why it exists.
+
+4. **Review the Code**
+
+   Read the code of the pull request and check it against some common criteria:
```
Comment (lsmith77) [inline contributing/community/reviews.rst]: just note that we will be able to (partially) automate some of these aspects eventually ..

File: contributing/community/reviews.rst [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+     edge cases?
+   * Does the PR contain sufficient comments to easily understand its code?
+   * Does the code break backwards compatibility? If yes, does the PR header say
+     so?
+   * Does the PR contain deprecations? If yes, does the PR header say so? Does
+     the code contain ``trigger_error()`` statements for all deprecated
+     features?
+   * Are all deprecations and backwards compatibility breaks documented in the
+     latest UPGRADE-X.X.md file? Do those explanations contain "Before"/"After"
+     with a clear upgrade path?
```
Comment (cordoval) [inline contributing/community/reviews.rst]: upgrade instructions?

File: contributing/community/reviews.rst [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+
+5. **Test the Code**
+
+   Take your project from step 2 and test whether the PR works properly.
```
Comment (lsmith77) [inline contributing/community/reviews.rst]: I guess step 3 (ie. Reproduce the Problem)?

File: contributing/community/reviews.rst [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+
+.. topic:: Example
+
+    Here is a sample comment for a PR that is not yet ready for merge:
+
+    .. code-block:: text
+
+        Thank you @weaverryan for working on this! It seems that your test
+        cases don't cover the cases when the counter is zero or smaller.
+        Could you add some tests for that?
```
Comment (cordoval) [inline contributing/community/reviews.rst]: Could you please add ...

**Comment by lsmith77** [comment]
Looks like a good step towards reducing the work for core developers and providing a breeding ground for new core developers!

File: contributing/community/reviews.rst (hunk 4 above)
Comment (webmozart) [inline contributing/community/reviews.rst]: From my experience there is almost no bug that is easier to reproduce without a SE fork than with. Even if a bug sounds simple, you need to set up the environment, test whether the bug can be reproduced, if not (often the case) understand why etc. I personally have switched to asking for SE forks even for simple bugs. This way the risk of wasting time is much lower.

File: contributing/community/reviews.rst (hunk 15 above)
Comment (webmozart) [inline contributing/community/reviews.rst]: I think the current example provides less friction as you can copy paste the repo URL from the browser, whereas you need to some more clicks to find the URL of the correct branch.

**Comment by webmozart** [comment]
Thanks for the feedback! I updated the PR now.

File: contributing/community/reviews.rst (hunk 14 above)
Comment (webmozart) [inline contributing/community/reviews.rst]: How can a bug report be a new feature? Shouldn't we rather update "Bug" to "Feature" then and remove the issue from he review process?

File: contributing/community/reviews.rst [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+     latest UPGRADE-X.X.md file? Do those explanations contain "Before"/"After"
+     examples with clear upgrade instructions?
+
+   Eventually, some of these aspects will be checked automatically.
+
+#. **Test the Code**
+
+   Take your project from step 3 and test whether the PR works properly.
+
+   TODO: precise steps
```
Comment (weaverryan) [inline contributing/community/reviews.rst]: Probably we don't really need to say much more here anyways. We can probably just remove this TODO line

**Comment by weaverryan** [comment]
There's one little `TODO` line floating in there that needs to be removed. But :+1: from me. @webmozart if you're ready for this to be merged - just say so, I can easily remove that TODO line for you.

**Comment by stof** [comment]
@webmozart could you finish this one ? It would be great to have this merged and a blog post published so that the experiment could start for real

File: contributing/community/reviews.rst (hunk 21 above)
Comment (webmozart) [inline contributing/community/reviews.rst]: I think for an average user there's much more to say. Sure, for us it's obvious, but the goal is to lower the barrier to entrance so more guidance = better IMO

**Comment by webmozart** [comment]
I just added the test instructions. It wasn't much after all. If everything is fine by you, this is ready to merge.

File: contributing/community/reviews.rst [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+#. **Test the Code**
+
+   Take your project from step 3 and test whether the PR works properly.
+   Replace the Symfony vendor by the code in the PR by running the following Git
+   commands. Insert the PR ID for the ``<ID>`` placeholders:
+
+   .. code-block:: text
+
+       $ cd vendor/symfony/symfony
+       $ git fetch origin pull/<ID>/head:pr<ID>
```
Comment (WouterJ) [inline contributing/community/reviews.rst]: is it really called origin? Afaik, Composer calls its repository `composer`

Comment (WouterJ) [inline contributing/community/reviews.rst]: Hmm, just checked and origin is available (besides composer).

**Comment by webmozart** [comment]
This might be a useful resource, also for a more generic guide for contributing patches: http://hoa-project.net/En/Literature/Contributor/Guide.html#Testing_a_patch

**Comment by weaverryan** [comment]
I just merged this - blog post coming next week. Thanks Bernhard!

File: contributing/community/reviews.rst (hunk 4 above)
Comment (weaverryan) [inline contributing/community/reviews.rst]: I shared the same thought as @lsmith77 and @xabbuh, but I think that forks are really done too little, and we should experiment with trying to encourage them. Yes, it's more work for contributors, but it may also be more satisfying for them to ship some code with the bug in it, and ultimately, this is an effort to relieve burden on the core team (and a fork is the best way). Let's try it this way for now.

**Comment by webmozart** [comment]
@weaverryan Thanks! I just added two more commits to improve the page a little: #5660

**Comment by stof** [comment]
@weaverryan any news of the blog post ?

**Comment by weaverryan** [comment]
@stof I'll post it this afternoon actually - I had it last week, but early week is better. Here's the draft - if anyone has time and sees mistakes, a comment is warmly appreciated!

https://gist.github.com/weaverryan/b78487c79ae0ccf59121

**Comment by stof** [comment]
@weaverryan given that both the doc page and the blog post are linking to the Carsonbot repo, it might be a good idea to improve the readme a bit to explain what the bot is.

**Comment by weaverryan** [comment]
@stof Good suggestion - I've just done that. And I've just posted the blog - some traveling made me loose track of doing it yesterday :) http://symfony.com/blog/calling-for-issue-triagers-a-new-workflow-and-the-carson-butler
